libbeat/common/reload/reload.go

```diff
@@ -0,0 +1,123 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package reload
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+// Register holds a registry of reloadable objects
+var Register = newRegistry()
+
+// ConfigWithMeta holds a pair of common.Config and optional metadata for it
+type ConfigWithMeta struct {
+	// Config to store
+	Config *common.Config
+
+	// Meta data related to this config
+	Meta *common.MapStrPointer
+}
+
+// ReloadableList provides a method to reload the configuration of a list of entities
+type ReloadableList interface {
+	Reload(configs []*ConfigWithMeta) error
+}
+
+// Reloadable provides a method to reload the configuration of an entity
+type Reloadable interface {
+	Reload(config *ConfigWithMeta) error
+}
+
+type registry struct {
+	sync.RWMutex
+	confsLists map[string]ReloadableList
+	confs      map[string]Reloadable
+}
+
+func newRegistry() *registry {
+	return &registry{
+		confsLists: make(map[string]ReloadableList),
+		confs:      make(map[string]Reloadable),
```

Comment:
I don't expect entities sharing the name, these should map to a config key, like `metricbeat.modules` or `output`.

The problem with having a single definition is that then, the objects have to handle possible errors in the `Reload` call. Like for instance, having the `output` object receiving a list of configs. This split is semantic, as the two entities behave differently.

Reply:
The suggestion from @jsoriano has some values, because currently we could call `Register` and `RegisterList` with the same name (same map key). Having a single interface for both concept would mean that we only have one map.